File: 1C/exercises/credit.c
#include <stdio.h>
#include <stdbool.h>
#include <cs50.h>
#include <math.h>
#include <string.h>



int main(void) {
    long creditCard = get_long("Type the credit card number: ");

    // checks if the card is valid via it's length
    bool isCardValid = true;
    int length = 0;
    if (creditCard >= pow(10, 15) && creditCard < pow(10, 16)) {
        length = 16;
    }
    else if (creditCard >= pow(10, 14) && creditCard < pow(10, 15) ) {
        length = 15;
    }
    else if (creditCard >= pow(10, 12) && creditCard < pow(10, 13)) {
        length = 13;
    }
    else {
        isCardValid = false;
    }

    // checks first digits to determine the card type later
    int firstDigit = creditCard / (long)pow(10, length-1);
    int firstTwoDigits = creditCard / (long)pow(10, length-2);

    // checks card type and validity via size and first digits
    string cardType = "";
    if (isCardValid && length == 16 && firstTwoDigits > 50 && firstTwoDigits < 56) {
        cardType = "MASTERCARD";
    }
    else if (isCardValid && length == 15 && (firstTwoDigits == 34 || firstTwoDigits == 37)) {
        cardType = "AMERICANEXPRESS";
    }
    else if (isCardValid && (length == 13 || length == 16) && firstDigit == 4) {
        cardType = "VISA";
    }
    else {
        isCardValid = false;
    }

    // Checks if card is valid via math
    if (isCardValid) {
        int oddDigit;
        int sumEvenDigits = 0;
        int multDigits = 0;
        for (int i = 1; i < length+1; i++) {
            if (i % 2 != 0) {
                oddDigit = (int)(creditCard / (long)pow(10, length-i) % 10);
                if (oddDigit * 2 > 10) {
                    multDigits = multDigits + (oddDigit * 2 / 10) + (oddDigit * 2 % 10);
                }
                else {
                    multDigits = multDigits + (oddDigit * 2);
                }
            }
            else {
                sumEvenDigits = sumEvenDigits + (int)(creditCard / (long)pow(10, length-i) % 10);
            }
        }

        if ((multDigits + sumEvenDigits) % 10 != 0) {
            isCardValid = false;
        }
    }

    if (isCardValid) {
        printf("%s\n", cardType);
    }
    else {
        printf("INVALID\n");
    }


    return 0;
}

File: 6Python/exercises/credit.py
from math import trunc

cardNumber = int(input("Number: "))

isCardValid = True
cardLength = len(str(cardNumber))
firstTwoDigits = trunc(cardNumber / (10 ** cardLength) * 100)
firstDigit = trunc(firstTwoDigits / 10)

if (cardLength == 13 or cardLength == 16) and firstDigit == 4:
    cardType = "VISA"
elif cardLength == 15 and (firstTwoDigits == 34 or firstTwoDigits == 37):
    cardType = "AMEX"
elif cardLength == 16 and (firstTwoDigits >= 51 and firstTwoDigits <= 55):
    cardType = "MASTERCARD"
else:
    isCardValid = False

summedNumbers = 0
multipliedNumbers = 0
currentMultipliedNumber = 0
for i in range(cardLength):
    if i % 2 != 0:
        currentMultipliedNumber = trunc(cardNumber / 10 ** i) % 10 * 2

        if currentMultipliedNumber > 9:
            multipliedNumbers += (currentMultipliedNumber % 10) + (trunc(currentMultipliedNumber / 10))
        else:
            multipliedNumbers += currentMultipliedNumber

    else:
        summedNumbers += trunc(cardNumber / 10 ** i) % 10
finalCardVerify = multipliedNumbers + summedNumbers

if isCardValid and finalCardVerify % 10 == 0:
    print(cardType)
else:
    print("INVALID")